PhysicsTools/PatUtils/plugins/L1PrefiringWeightProducer.cc

```diff
@@ -0,0 +1,478 @@
+
+class L1PrefiringWeightProducer : public edm::stream::EDProducer<> {
+public:
+  explicit L1PrefiringWeightProducer(const edm::ParameterSet&);
+  ~L1PrefiringWeightProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  double getPrefiringRate(double eta, double pt, TH2F* h_prefmap, fluctuations fluctuation) const;
+  double getPrefiringRateMuon(double eta, double phi, double pt, fluctuations fluctuation) const;
+
+  edm::EDGetTokenT<std::vector<pat::Photon> > photons_token_;
+  edm::EDGetTokenT<std::vector<pat::Jet> > jets_token_;
+  edm::EDGetTokenT<std::vector<pat::Muon> > muon_token_;
+
+  TF1* parametrization0p0To0p2_;
+  TF1* parametrization0p2To0p3_;
+  TF1* parametrization0p3To0p55_;
+  TF1* parametrization0p55To0p83_;
+  TF1* parametrization0p83To1p24_;
+  TF1* parametrization1p24To1p4_;
+  TF1* parametrization1p4To1p6_;
+  TF1* parametrization1p6To1p8_;
+  TF1* parametrization1p8To2p1_;
+  TF1* parametrization2p1To2p25_;
+  TF1* parametrization2p25To2p4_;
+  TF1* parametrizationHotSpot_;
+
+  TH2F* h_prefmap_photon;
+  TH2F* h_prefmap_jet;
+  std::string dataera_;
+  std::string dataeraMuon_;
+  bool useEMpt_;
+  double prefiringRateSystUnc_;
+  double jetMaxMuonFraction_;
+  bool skipwarnings_;
+};
+
+L1PrefiringWeightProducer::L1PrefiringWeightProducer(const edm::ParameterSet& iConfig) {
+  photons_token_ = consumes<std::vector<pat::Photon> >(iConfig.getParameter<edm::InputTag>("ThePhotons"));
+  jets_token_ = consumes<std::vector<pat::Jet> >(iConfig.getParameter<edm::InputTag>("TheJets"));
+  muon_token_ = consumes<std::vector<pat::Muon> >(iConfig.getParameter<edm::InputTag>("TheMuons"));
+
+  dataera_ = iConfig.getParameter<std::string>("DataEraECAL");
+  dataeraMuon_ = iConfig.getParameter<std::string>("DataEraMuon");
+  useEMpt_ = iConfig.getParameter<bool>("UseJetEMPt");
+  prefiringRateSystUnc_ = iConfig.getParameter<double>("PrefiringRateSystematicUncty");
+  jetMaxMuonFraction_ = iConfig.getParameter<double>("JetMaxMuonFraction");
+  skipwarnings_ = iConfig.getParameter<bool>("SkipWarnings");
+
+  TFile* file_prefiringmaps_;
+  std::string fname = iConfig.getParameter<std::string>("L1Maps");
+  edm::FileInPath mapsfilepath("PhysicsTools/PatUtils/data/" + fname);
+  file_prefiringmaps_ = new TFile(mapsfilepath.fullPath().c_str(), "read");
+  if (file_prefiringmaps_ == nullptr && !skipwarnings_)
+    edm::LogWarning("L1PrefireWeightProducer")
+        << "File with maps not found. All prefiring weights set to 0. " << std::endl;
+
+  TString mapphotonfullname = "L1prefiring_photonptvseta_" + dataera_;
+  if (!file_prefiringmaps_->Get(mapphotonfullname) && !skipwarnings_)
+    edm::LogWarning("L1PrefireWeightProducer")
+        << "Photon map not found. All photons prefiring weights set to 0. " << std::endl;
+  h_prefmap_photon = (TH2F*)file_prefiringmaps_->Get(mapphotonfullname);
```

Comment:
The (very coarse) type cast could be avoided with
```suggestion
  h_prefmap_photon = file_prefiringmaps_->Get<TH2F>(mapphotonfullname);
```

Reply:
Done.